### ISSUE TITLE
[analyzer_plugin] Running build_runner does not invoke `analysis.updateContent` with the newly generated files in VScode

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/sdk#54113]
Hello!

It appears that when generated files are generated using build_runner, analyzer_plugins do not receive a `analysis.updateContent` matching the generation.

Steps to reproduce:

- open a project using build_runner and an analyzer_plugin
- run code-generation once
- delete a .g.dart (this will correctly invoke (analysis.updateContent with `{type: remove}`)
- run `build_runner build -d`
- The generated `.g.dart` file will be re-created. But no `analysis.updateContent` with `{type: add}` will be sent

This means that subsequent `getFixes`/`getsAssists` behave as if code-generation did not happen.

Here's a video showcasing the problem.

- First, I deleted `dependencies.g.dart`. This correctly updates lints due to missing generated files.  
  We see in the bottom of the screen at that point that an `updateContent` was sent, with the content of the generated file.
- Then, I ran `dart run build_runner build -d`.
  The `dependencies.g.dart` reappeared. But as you can see in the logs, no `updateContent` request was sent. So the IDE behaves as if the generated file is still missing.
- Opening the generated file fixes the issue as it appears to forcibly call `updateContent`.

https://github.com/dart-lang/sdk/assets/20165741/3f9aeaf2-1b0d-4554-a424-c864409dc5e1

**Comment by rrousselGit** [comment]
It appears that instead of deleting the `.g.dart` in the IDE, we call `mv dependencies.g.dart backup.dart`, then the `{type: remove}` event isn't sent either. 

It may be that for all file changes triggered by the terminal, analyzer_plugin isn't notified.

**Comment by bwilkerson** [comment]
Definitely something we will need to get right in the new plugin implementation.

@srawlins

**Comment by bwilkerson** [comment]
We're in the design phase, and once we have a design we can make a decision about whether we will implement it.

**Comment by bwilkerson** [comment]
The `analysis.handleWatchEvents` request is sent to the plugins from `PluginManager.broadcastWatchEvent`
- which is invoked from `AnalysisServer.broadcastWatchEvent`,
- which is invoked, indirectly, from [a callback in `ContextManagerImpl._createAnalysisContexts`](https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/context_manager.dart#L572).

I don't know how much that helps, though.

Note that our use of the watcher package is wrapped so that we can use it from the `ResourceProvider`. That allows us to write in-memory tests that simulate the behavior of the watcher package without having to actually modify the disk. It might make it easier to debug the issue.

**Comment by rrousselGit** [comment]
About this, is there a document detailing how to use the local clone of `analyzer_server` in my IDE?

I have tests running and was looking into forking `test_concurrentContextRebuilds` to debug this. But so far I'm struggling to write a test for this, and would likely need a debugger first.

**Comment by rrousselGit** [comment]
I don't really have time to work on this, as you've seen. But it's a pretty important bug.

I see users complain about this about every week.

**Comment by DanTup** [comment]
@rrousselGit I was just investigating a similar issue about watch events not getting to the plugin (custom_lint) and I think it's a simple fix in the plugin - see https://github.com/invertase/dart_custom_lint/issues/330#issuecomment-2903900798

Is this likely the same issue, or something different? If something different, do you have steps to reproduce (it's not clear to me if it's custom_lint or something else in your video).


> About this, is there a document detailing how to use the local clone of analyzer_server in my IDE?

If you're using VS Code, you can set up the SDK (see https://github.com/dart-lang/sdk/blob/main/docs/Building.md#getting-the-source) and then set the `dart.analyzerPath` setting to run the server from source:

```
"dart.analyzerPath": "D:\\Dev\\Google\\dart-sdk\\sdk\\pkg\\analysis_server\\bin\\server.dart",
```

It'll be a little slower to start up, but then you can make changes to the analysis server source and run the **Reload Window** command in VS Code to restart the server.

**Comment by rrousselGit** [comment]
I'll have to double check, but that sounds an interesting call out.

But if that's the case, I don't see why .dart files work but not .*.dart ones, as both use the same glob and are located in the same folder

**Comment by DanTup** [comment]
> But if that's the case, I don't see why .dart files work but not .*.dart ones

Oh, I didn't realise that was a difference - is it definitely this, and not just which files you do/don't have open in the IDE? The issue I was debugging was that the watch events from disk were not being passed over, but the overlays (which exist for files open in the IDE) are updated differently.

If you have steps to repro, I can do some more debugging though (or you can apply the fix for https://github.com/invertase/dart_custom_lint/issues/330 and test whether that changes anything).

**Comment by rrousselGit** [comment]
>  is it definitely this, and not just which files you do/don't have open in the IDE?

Yes, I'm positive about it. I regularly see lints appears in files that are not open in the IDE. In fact, my whole testing flow for lint rules relies on that, so I'm quite confident that this works.    
It feels specific to writes coming from the terminal ; as opposed to writes from the IDE.


Although I'll test your possible fix asap to confirm this

**Comment by DanTup** [comment]
> Yes, I'm positive about it. I regularly see lints appears in files that are not open in the IDE. In fact, my whole testing flow for lint rules relies on that, so I'm quite confident that this works.

To be clear, I'm talking about watch events only. The issue with https://github.com/invertase/dart_custom_lint/issues/330 is that the files are linted correctly at startup, but if the file is modified on-disk (eg. not by being opened and changed in the IDE) that the watch event is not sent to the plugin (because of the bad glob) and therefore the lines don't _update_ afterwards.

**Comment by rrousselGit** [comment]
I didn't think about that.  
That sounds plausible. I'll double check.

**Comment by srawlins** [comment]
We've deprecated the legacy plugin system, as of Dart 3.14.2, and will be removing support for the legacy plugin system in an upcoming release of the Dart SDK. The [new plugins system](https://dart.dev/tools/analyzer-plugins) is supported and continually improving.
